FK_Project/src/Fisher_Kolmogorov.hpp

```diff
@@ -0,0 +1,252 @@
+#ifndef FISHER_KOLMOGOROV_HPP
+#define FISHER_KOLMOGOROV_HPP
+
+#include <deal.II/base/conditional_ostream.h>
+#include <deal.II/base/quadrature_lib.h>
+
+#include <deal.II/distributed/fully_distributed_tria.h>
+
+#include <deal.II/dofs/dof_handler.h>
+#include <deal.II/dofs/dof_tools.h>
+
+#include <deal.II/fe/fe_simplex_p.h>
+#include <deal.II/fe/fe_system.h>
+#include <deal.II/fe/fe_values.h>
+#include <deal.II/fe/fe_values_extractors.h>
+#include <deal.II/fe/mapping_fe.h>
+
+#include <deal.II/grid/grid_in.h>
+#include <deal.II/grid/grid_out.h>
+#include <deal.II/grid/tria.h>
+
+#include <deal.II/lac/solver_cg.h>
+#include <deal.II/lac/trilinos_precondition.h>
+#include <deal.II/lac/trilinos_sparse_matrix.h>
+#include <deal.II/lac/vector.h>
+
+#include <deal.II/numerics/data_out.h>
+#include <deal.II/numerics/matrix_tools.h>
+#include <deal.II/numerics/vector_tools.h>
+
+#include <fstream>
+#include <iostream>
+
+using namespace dealii;
+
+class Fisher_Kolmogorov
+{
+public:
+
+    // Physical dimension (1D, 2D, 3D)
+    static constexpr unsigned int dim = 3;
+
+    // Function describing the value of the spreading coefficient -> the bigger alpha is, the faster is the spreading
+    class FunctionAlpha : public Function<dim> 
+    {
+    public : 
+        virtual double 
+        value(const Point<dim> & /*p*/,
+            const unsigned int /*component*/ = 0) const override
+        {   
+            return 0.5;
+        }
+    }; 
+
+    // Function describing the behaviour of D
+    class FunctionD : public Function<dim> 
+    {
+    public : 
+        virtual void 
+        matrix_value(const Point<dim> & /*p*/,
+            FullMatrix<double> &values) const /*override*/
+        {   
+            // Here go through the diagonal the values for the extracellular diffusion
+            for(unsigned int i = 0; i < dim; i++ ){
+                for(unsigned int j = 0; j < dim; j++){
+                    if(i == j){
+                        //da ricontrollare 
+                        values(i,j) = 1.0; 
+                    }
+                    else{
+                        values(i,j) = 0.0; 
+                    }
+                }
+            }
+
+            // Here should go the values for the d_axn
+            // Example: values(1,1) += 3.0;
+        }
+
+        virtual double 
+        value(const Point<dim> &/*p*/, const unsigned int component1 = 0, const unsigned int component2 = 1)  const /*override*/ 
+        {
+            return (component1 == component2) ? 1.0 : 0.0;
+        }
+    }; 
```

Comment:
Mathematically speaking, $D$ is a tensor-valued field, i.e. a function of space that returns a rank-2 tensor. The best representation you can give in code is that of a `Tensor<2,dim>`-valued function. `deal.ii` has a `TensorFunction<rank, dim>` class for this purpose. Alternatively, you might consider using the `Function<dim>` class with `dim*dim` components (mapping the matrix elements into a linear ordering).

Either way, you should **never** change the interface of a class you inherit from. For instance, the `Function<dim>` class has a method `double value(const Point<dim> &, const unsigned int)` that evaluates the function. Here you redefine your own `value` method, that has a different signature (different arguments) and therefore does **not** override `Function<dim>::value`, but rather it overloads it (which, I assume, is the reason why you commented out the `override` keyword).

Now, consider the following situation. Some user of your code (possibly your future selves) wants to interpolate the value of this function on all nodes of a mesh, for some reason. To do so, they call `VectorTools::interpolate`: after all, this object is a `Function<dim>`, it is reasonable to expect it to behave as such (that is the meaning of class inheritance), and so it is reasonable to believe I can feed it to any function that takes a `Function<dim>` in input.

However, `VectorTools::interpolate` calls `value(const Point<dim>&, const unsigned int)`, that is the version of the method present in the interface of `Function<dim>`. I'm not sure what the `Function<dim>::value` method does by default, but I'm pretty sure it does not do the same as your overloaded version. Maybe it throws an exception, maybe it returns zero without giving any warning. In the latter case, you have a very hard to find bug (especially if the user of the code does not know or remember the design choices made in the implementation of `FunctionD`).

tl;dr: never depart from the expected interface of a parent class unless you have a very very very strong reason to do so and you trust your software design skills very very very much. Otherwise, you should either not inherit from that class, or change your code to work with its established interface.